python/cuml/cuml/internals/interop.py

```diff
@@ -0,0 +1,311 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import annotations
+
+import functools
+import warnings
+from importlib import import_module
+from typing import Any
+
+from cuml.internals.device_type import DeviceType
+from cuml.internals.global_settings import GlobalSettings
+from cuml.internals.mem_type import MemoryType
+from cuml.internals.memory_utils import using_output_type
+
+__all__ = (
+    "UnsupportedOnGPU",
+    "UnsupportedOnCPU",
+    "InteropMixin",
+    "warn_legacy_device_interop",
+    "to_gpu",
+)
+
+
+def warn_legacy_device_interop(gpu_func):
+    """A decorator for warning users that device selection no longer
+    works on methods previously decorated by `enable_device_selection`."""
+
+    @functools.wraps(gpu_func)
+    def inner(self, *args, **kwargs):
+        cls = type(self)
+        gpu_cls_name = cls.__name__
+        cpu_cls_path = cls._cpu_class_path
+        if GlobalSettings().device_type == DeviceType.host:
+            from cuml.common.device_selection import using_device_type
+
+            warnings.warn(
+                f"Support for setting the `device_type` to execute `{gpu_cls_name}` "
+                f"methods on CPU was removed in version 25.06. Please use "
+                f"`{cpu_cls_path}` directly to handle CPU execution, and "
+                f"`{gpu_cls_name}.from_sklearn`/`{gpu_cls_name}.as_sklearn` "
+                f"to manage conversion to/from cuML as needed."
+            )
+            # Some code internally still checks device_type and errors on the GPU
+            # path if device_type isn't device. For now we warn if set to host
+            # and change to device temporarily for just this method.
+            with using_device_type("device"):
+                return gpu_func(self, *args, **kwargs)
+        else:
+            return gpu_func(self, *args, **kwargs)
+
+    return inner
+
+
+def to_gpu(x, order="K"):
+    """Coerce `x` to the equivalent GPU type."""
+    # Ideally we'd get rid of this helper eventually, using it for now to reduce typing
+    from cuml.internals.input_utils import input_to_cuml_array
+
+    return input_to_cuml_array(
+        x, order=order, convert_to_mem_type=MemoryType.device
+    )[0]
+
+
+class UnsupportedOnGPU(ValueError):
+    """An exception raised when a conversion of a CPU to a GPU estimator isn't supported"""
+
+
+class UnsupportedOnCPU(ValueError):
+    """An exception raised when a conversion of a GPU to a CPU estimator isn't supported"""
+
+
+class classproperty:
+    """Like ``property``, but on a class rather than an instance."""
+
+    def __init__(self, fget):
+        self.fget = fget
+
+    def __get__(self, obj, owner):
+        return self.fget(owner)
+
+
+class InteropMixin:
```

Comment:
I thought we were removing all the machinery for the old style "train on GPU, infer on CPU" dispatching. Which means I wasn't expecting to see the `InteropMixin` which is used by cuml classes. I was expecting all code related to CPU/GPU conversion to live within the accelerator and have nothing in the actual cuml code related to converting back and forth.

If we keep this, can't we also keep the context manager and such? Or do we only need this to enable `as_sklearn`/`from_sklearn`?